### ISSUE TITLE
Container App/Container App Environment outbound IP limitations

### ISSUE THREAD

**Issue by ErikRosengren-work** [microsoft/azure-container-apps#801]
### This issue is a: (mark with an x)
- [ ] bug report -> please search issues before submitting
- [x] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
The following has recently been added to the documentation describing networking for container apps. 

> **Outbound public IP**: Used as the "from" IP for outbound connections that leave the virtual network. These connections aren't routed down a VPN. **Outbound IPs aren't guaranteed and may change over time**. Using a NAT gateway or other proxy for outbound traffic from a Container App environment is only supported on the workload profile environment.

Quote from: https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/container-apps/networking.md?plain=1#L106

I need clarification here regarding exactly what this means. Does this mean that the outbound IP that the container app gets from the container app environment can change? Because right now, as far as I can tell, the outbound IP-address of a container app in a managed container environment is determined by one of the IP addresses that can be found in a separate resource group with a generated name. 

The reason I ask is because in my use case I need to guarantee that the outbound IP addresses don't change because of whitelisting.

**Comment by torosent** [comment]
The outbound IP can change since there is no guarantee from the underlying infrastructure to have to same IP. The only way to guarantee it is to add a NAT gateway to your subnet or any other egress control appliance like Azure Firewall. 
Note that this solution applies for Workload Profiles (Consumption + Dedicated) mode. You cannot use NAT gateway in Consumption only mode.

**Comment by fritzfs** [comment]
@torosent thank you. 

So, [this comment](https://github.com/microsoft/azure-container-apps/issues/355#issuecomment-1300619829) is not relevant anymore? 

Because, at this moment, I see approx 30 outbound IP addresses for my ACA with ACA Environment using Workload Profiles and Consumption plan and custom VNET. 

Another question. So, I can configure NAT Gateway if I use Consumption plan in Workload Profile?

**Comment by Yuvraj102** [comment]
Hello Did you guys find any solution for this ?, I need static outbound ip for my ACA as well

**Comment by wUEZRs** [comment]
I will note down for the history that we ran a Container App on Consumption and had NAT Gateway enabled for a month on its subnets. After a 26 days we got a lot of network errors "out of the blue"

We could not handle any of the containers, with a lot of different InternalServerErrors, seemingly also in the Azure portal.

```
Stop-AzContainerApp -Namexxx -ResourceGroupNamexxxx -SubscriptionId xxxxxx
Stop-AzContainerApp_Stop: Failed to stop container app 'xxxxx-frontend-app'. Error details: Operation returned an invalid status code 'InternalServerError', response body {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Internal error occurred: failed calling webhook \"mcontainerapp.kb.io\": failed to call webhook: Post \"https://k8se-containerapp-webhooks-service.k8se-system.svc:443/mutate-k8se-microsoft-com-v1alpha1-containerapp?timeout=10s\": no endpoints available for service \"k8se-containerapp-webhooks-service\"","reason":"InternalError","details":{"causes":[{"message":"failed calling webhook \"mcontainerapp.kb.io\": failed to call webhook: Post \"https://k8se-containerapp-webhooks-service.k8se-system.svc:443/mutate-k8se-microsoft-com-v1alpha1-containerapp?timeout=10s\": no endpoints available for service \"k8se-containerapp-webhooks-service\""}]},"code":500}
```

I think this means that if you enable it, it will mess up your environment after a while 😊
And that we "can" technically use a NAT gateway, and Azure does not stop you from doing it, with the consumption plan. It stops you from say an SQL MI, but here it did not in its deployment.

**Comment by n0ngalaxy** [comment]
Hi all, I'm quite new to this and am confused. Please correct me if I'm wrong. 😅 

So, in short, to have a static outbound IP address for ACA, we'll need to configure a workload profile for ACA, instead of using a consumption-only plan? And we'll need to configure the VNet and subnet during ACA creation, and attach them to a NAT Gateway???

**Comment by jjsridharan** [comment]
### ⚠️ NAT Gateway + Azure Container Apps: important gotcha

If you’re trying to use a **NAT Gateway with Azure Container Apps**, be aware of this:

> **NAT Gateway does NOT work reliably with a *Consumption-only* Container Apps environment.**

I spent a fair amount of time trying different workarounds to make it work (routing tweaks, deployment ordering, infra debugging, etc.). None of them were worth it. As another community member (wUEZRs) pointed out, *you might be able to force it to work in some edge cases, but it’s fragile and you’re very likely to run into issues later*.

This limitation is **actually documented**, though it’s easy to miss.

Below excerpt from the official docs:
[https://learn.microsoft.com/en-us/azure/container-apps/networking?source=recommendations](https://learn.microsoft.com/en-us/azure/container-apps/networking?source=recommendations)

<img width="937" height="483" alt="Image" src="https://github.com/user-attachments/assets/37fc3ec3-afa1-4113-90a6-73bf2a3ae5f9" />

And from the custom VNET docs (Consumption-only environment):
[https://learn.microsoft.com/en-us/azure/container-apps/custom-virtual-networks?tabs=consumption-only-env](https://learn.microsoft.com/en-us/azure/container-apps/custom-virtual-networks?tabs=consumption-only-env)

<img width="925" height="381" alt="Image" src="https://github.com/user-attachments/assets/9e4da38e-4258-4d78-84ce-b32e3e0d759c" />

The key takeaway from the docs:

* **Consumption-only environment** ❌ does **not** support custom egress (NAT Gateway, UDRs, etc.)
* **Workload Profile environment** ✅ *does* support NAT Gateway and enterprise networking scenarios

---

### ✅ The fix: switch to a Workload Profile environment

If you’re using **Bicep**, the good news is that the fix is very small.

A **Workload Profile environment** supports **two workload types**:

* `Consumption`
* `Dedicated`

You can still choose **Consumption** inside a Workload Profile environment, which effectively gives you:

* Serverless-style billing
* **AND** proper support for NAT Gateway / VNET / UDRs

In other words, you get *the best of both worlds*.

Below (see image) are the **only changes required** to switch from a Consumption-only environment to a Workload Profile environment.

<img width="1203" height="660" alt="Image" src="https://github.com/user-attachments/assets/bf7ff319-9a86-41ac-9395-6d7caee522f1" />

After making this switch:

* NAT Gateway integration worked immediately
* Environment provisioning completed cleanly
* No more stuck `Waiting` state or infra LB health probe issues

---

### TL;DR

* ❌ NAT Gateway + **Consumption-only** Container Apps = pain
* ✅ NAT Gateway + **Workload Profile environment** = supported and stable
* You can still run apps on **Consumption** inside a Workload Profile env
* If you need static egress IPs or enterprise networking, **start with Workload Profiles**

If others have additional nuances or cost/perf considerations, please add them — but this environment switch alone unblocked everything for me and aligns with the official docs.

Hope this saves someone else a few hours (or days).
